batch/batch/server/database.py

```diff
@@ -0,0 +1,167 @@
+                                               db=self.db,
+                                               user=self.user,
+                                               password=self.password,
+                                               charset=self.charset,
+                                               cursorclass=aiomysql.cursors.DictCursor,
+                                               autocommit=True)
+
+    async def temp_table_name(self, root):
+        suffix = uuid.uuid4().hex[:8]
+        name = f'{root}-{suffix}'
+        niter = 0
+        while await self.has_table(name):
+            suffix = uuid.uuid4().hex[:8]
+            name = f'{root}-{suffix}'
+            niter += 1
+            if niter > 5:
+                raise Exception("Too many attempts to get unique temp table.")
+        return name
+
+    def temp_table_name_sync(self, root):
+        return run_synchronous(self.temp_table_name(root))
+
+    async def has_table(self, name):
+        async with self.pool.acquire() as conn:
+            async with conn.cursor() as cursor:
+                sql = f"SELECT * FROM INFORMATION_SCHEMA.tables " \
+                    f"WHERE TABLE_SCHEMA = %s AND TABLE_NAME = %s"
+                await cursor.execute(sql, (self.db, name))
+                result = cursor.fetchone()
+        return result.result() is not None
+
+    def has_table_sync(self, name):
+        return run_synchronous(self.has_table(name))
+
+    async def drop_table(self, *names):
+        async with self.pool.acquire() as conn:
+            async with conn.cursor() as cursor:
+                await cursor.execute("DROP TABLE IF EXISTS {}".format(",".join([f'`{name}`' for name in names])))
+
+    def drop_table_sync(self, *names):
+        return run_synchronous(self.drop_table(*names))
+
+    async def create_table(self, name, schema, keys):
+        assert all([k in schema for k in keys])
+
+        async with self.pool.acquire() as conn:
+            async with conn.cursor() as cursor:
+                schema = ", ".join([f"`{n}` {t}" for n, t in schema.items()])
+                key_names = ", ".join([f'`{name.replace("`", "``")}`' for name in keys])
+                keys = f", PRIMARY KEY( {key_names} )" if keys else ''
+                sql = f"CREATE TABLE IF NOT EXISTS `{name}` ( {schema} {keys})"
+                await cursor.execute(sql)
+
+    def create_table_sync(self, name, schema, keys):
+        return run_synchronous(self.create_table(name, schema, keys))
+
+    async def create_temporary_table(self, name, schema, keys):
+        name = await self.temp_table_name(name)
+        t = await Table(self, name, schema, keys)
+        return t
+
+    def create_temporary_table_sync(self, name, schema, keys):
+        return run_synchronous(self.create_temporary_table(name, schema, keys))
+
+
+def make_where_statement(items):
+    template = []
+    values = []
+    for k, v in items.items():
+        if isinstance(v, list):
```

Comment:
I don't think this is necessary; [`IN` should be optimized for single values](https://stackoverflow.com/a/31754630).

Another ref: https://dba.stackexchange.com/questions/58940/what-is-the-performance-difference-between-in-42-and-id-42-in-mysql

If the optimizer failed the difference would be negligible.

Reply:
So is your proposal to only accept lists and not singular values? I don't like that as an interface. Then I'd have to add this which is essentially the same thing.

```
if not isinstance(v, list) and not isinstance(v, tuple):
    v = list(v)
```

Reply:
```python
def test(data):
    template = []
    values = []
    for k, v in data.items():
        if not v:
            template.append("FALSE")
        else:
            template.append(f'`{k.replace("`", "``")}` IN %s')
            values.append(v)

    return " AND ".join(template), values


if __name__ == "__main__":
    data = {"k1": "v1", "k2": "v2", "k3": [
        "v3a", "v3b", "v3c"], "key_false": [], "key_false2": ""}

    keys, values = test(data)

    print("keys", keys)
    print('values', values)
```

Also, did you mean to do?
```
if len(v) == 0:
                 template.append("FALSE")
```

This would do something like ``WHERE FALSE` which doesn't seem right, or maybe would be a no-op. If you wanted to do a null check, maybe you would do instead:

```
if not v:
  template.append(f'`{k.replace("`", "``")}` IS %s')
  values.append(None)
else:
  ...
```

Lastly, I'm not quite certain if %s will properly handle all cases. `%r` seems more correct, but the argument escapers built into pymysql and aiomysql (which seems to borrow the former's functionality) probably do the right thing, and I don't see `%r` being used, ever.

```sh
# For values (False, True, "hello", 1)
Test R1: (False, True, 'hello', 1)
Test S1: (False, True, hello, 1)
Test R2: ('a tuple', (1, 2, 3))
Test S2: (a tuple, (1, 2, 3))
```

Reply:
The issue is you have to do something if the list is empty. I see it as equivalent of `ID in ()` which is always False. https://stackoverflow.com/questions/13210233/empty-in-clause-parameter-list-in-mysql

The problem with your code is a value might actually be Null. In which case your code returns records when it shouldn't.

I don't see %r in any of the examples.

Reply:
Wait, why is the caller passing () as a value? What does that mean?

If I ask for values {"key1": ()} what am I asking for? I think this should be unsupported, or we should make a decision that this means the user is requesting that their column is empty (NULL). 

I would feel more comfortable with completely not supporting falsy values, and allowing the user to pass None as a value.

Reply:
I wanted something general and this is used in later PRs where I programmatically ask for the parent records of each job and it can be an empty list. So I can't throw an exception and I have to handle this use case.

Reply:
Ah. From that link, I gather the result in your case is always an empty set. I don't quite see the advantage of building the full result and incurring the network latency.

I would prefer for the caller to check whether a subquery could be built, and avoid the network latency when possible.

However, if you want to build a query, could you return as soon as you find an empty-set condition?

```py
if len(v) == 0:
  return "WHERE %s", (False)
```

Reply:
I bet the SQL compiler will short circuit and do the right thing. Can we make this an optimization for another PR in the future? This is working.

Reply:
Ok